[Mono]: Additional support for signal async safe stack-walks.

lateralusX opened the pull request:
Mono's stack-walker can be signal/async safe when needed, making sure it won't allocate additional memory or taking internal runtime locks. It is however up to the caller of the stack walking API's to decide if it should be signal and/or async safe. Currently this is controlled using two different flags, the `MONO_UNWIND_SIGNAL_SAFE` as well as `mono_thread_info_is_async_context`.

This is problematic since callers might want signal async safe stack-walking but since not both are always set, it will not behave fully signal async safe.

https://github.com/dotnet/android/issues/9365 hit a couple of scenarios described here:

https://github.com/dotnet/android/issues/9365#issuecomment-2725069793

that ends up deadlocking due to the fact that it did a stack-walking from within a signal handler and deadlocked when dumping stack on suspended thread holding runtime loader lock, but without making the stack-walk async safe.

PR makes sure that calls to stack-walk API's can be made signal and async safe by using MONO_UNWIND_SIGNAL_SAFE stack-walking flag. PR also makes sure stack-walks that deadlocked in https://github.com/dotnet/android/issues/9365#issuecomment-2725069793 uses this flag triggering a signal async safe stack-walk.

A signal safe stack-walk should in theory also be an async safe stack-walk (but not the other way around), so `MONO_UNWIND_SIGNAL_SAFE` should have both behaviors, signal and async safe, but current flag is old and was set to the same value as MONO_UNWIND_NONE and was used in a couple of scenarios that could potentially cause regression when hehavior changed to async safe mode.

This PR fix `MONO_UNWIND_SIGNAL_SAFE`, to be both signal and async safe and used where stack-walk needs to be real signal safe. In order to prevent regression of old code previously using MONO_UNWIND_SIGNAL_SAFE have been changed to use MONO_UNWIND_NONE to prevent any regression. MONO_UNWIND_SIGNAL_SAFE and MONO_UNWIND_NONE ended up as the same value, 0, so using MONO_UNWIND_NONE should preserve old behavior. If it turns out that these stack-walks needs to be fully signal safe in the future (they have not been so far), then they could be migrated over to use enhanced MONO_UNWIND_SIGNAL_SAFE.

lateralusX committed:
[Mono]: Fix additional stack-walks to be async safe.

Mono's stack-walker can be signal/async safe when needed,
making sure it won't allocate additional memory or taking internal
runtime locks. It is however up to the caller of the stack walking API's
to decide if it should be signal and/or async safe. Currently this is
controlled using two different flags, the MONO_UNWIND_SIGNAL_SAFE as well
as mono_thread_info_is_async_context.

This is problematic since callers wants signal safe stack-walking but
since not both are set, it will not behave fully signal safe.

https://github.com/dotnet/android/issues/9365 hit a couple of scenarios
described here:

https://github.com/dotnet/android/issues/9365#issuecomment-2725069793

that ends up deadlocking due to the fact that they did stack-walking
from within a signal handler and deadlocked or dumping stack on suspended
thread holding runtime loader lock, but without making the stack-walk async safe.

Fix makes sure that calls to stack-walk API's can be made signal and/or
async safe and that identified areas uses the correct set of flags given
state of threads when stack-walking.

lateralusX committed:
Add signal async safe stack unwind option.

lateralusX committed:
Assert that walk_stack_full_llvm_only is only called in llvm only mode.

lateralusX committed:
Correct some bool usage.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono-runtime-provider.c:
```diff
@@ -1184,31 +1184,27 @@ ep_rt_mono_walk_managed_stack_for_thread (
-	bool restore_async_context = FALSE;
-	bool prevent_profiler_event_recursion = FALSE;
+	bool prevent_profiler_event_recursion = false;
+	MonoUnwindOptions unwind_options = MONO_UNWIND_SIGNAL_SAFE;
```
For regular stack-walks, using MONO_UNWIND_SIGNAL_ASYNC_SAFE will mean that we won't try to load anything during stack-walk if not already loaded, so if we have code that used signal safe, then switching to signal async safe might end up with some unresolved frames if that scenario gets hit, but in that case, maybe that stack-walk code should not run in signal safe environments and if it doesn't need it, it should go with one of the other flags.

lateralusX replied:
OK, yes, we could do that, that should be without any regression risk.

lateralusX replied:
and for code that really needs signal safe, that includes async safe, then then new MONO_UNWIND_SIGNAL_ASYNC_SAFE should be used.

BrzVlad replied:
> or do you suggest that we should actually make the old MONO_UNWIND_SIGNAL_SAFE a real enum value and let it behave as MONO_UNWIND_SIGNAL_ASYNC_SAFE does in this PR? So what we end up with in the end, all current use of MONO_UNWIND_SIGNAL_SAFE -> MONO_UNWIND_NONE and then all use of MONO_UNWIND_SIGNAL_ASYNC_SAFE -> MONO_UNWIND_SIGNAL_SAFE.

I wasn't suggesting this, but I think it sounds just as good. However you feel it makes more sense, just not to have 2 SIGNAL_SAFE values, where only one works.

lateralusX committed:
Make signal safe unwind option, signal async safe.

Mono's current MONO_UNWIND_SIGNAL_SAFE was not fully signal safe since
it was not async safe, that could lead to taking loader lock. This
will fix MONO_UNWIND_SIGNAL_SAFE to be signal asycn safe, it will also
change current use of MONO_UNWIND_SIGNAL_SAFE to MONO_UNWIND_NONE since
they where equal before this fix, meaning old calls using
MONO_UNWIND_SIGNAL_SAFE will behave identical using MONO_UNWIND_NONE
so no regression.

lateralusX commented:
Errors in "Build browser-wasm linux Release" lanes are reported as known issues.

lateralusX commented:
@steveisok, backport?
